src/RedCode/Currency/Rate/Provider/YahooCurrencyRateProvider.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+namespace RedCode\Currency\Rate\Provider;
+
+use RedCode\Currency\ICurrency;
+use RedCode\Currency\ICurrencyManager;
+use RedCode\Currency\Rate\ICurrencyRate;
+use RedCode\Currency\Rate\ICurrencyRateManager;
+
+class YahooCurrencyRateProvider implements ICurrencyRateProvider
+{
+    const PROVIDER_NAME = 'yahoo';
+
+    public function __construct(ICurrencyRateManager $currencyRateManager, ICurrencyManager $currencyManager)
+    {
+        $this->currencyRateManager  = $currencyRateManager;
+        $this->currencyManager      = $currencyManager;
+    }
+
+    /**
+     * Load rates by date
+     *
+     * @param ICurrency[] $currencies
+     * @param \DateTime|null $date
+     *
+     * @return ICurrencyRate[]
+     */
+    public function getRates($currencies, \DateTime $date = null)
+    {
+        $currencyCodes = [];
+
+        foreach ($currencies as $currency) {
+            $currencyCodes[] = $currency->getCode();
+        }
+
+        if (null === $date) {
+            $date = new \DateTime();
+            $results = $this->_processActualRequest($currencyCodes);
+        } else {
+            $results = $this->_processHistoricalRequest($currencyCodes, $date);
+        }
+
+        $rates = [];
+        foreach ($results as $code => $rate) {
+            $rates[$code] = $this->currencyRateManager->getNewInstance(
+                $this->currencyManager->getCurrency($code),
+                $this,
+                $date,
+                $rate,
+                1
+            );
+        }
+
+        return $rates;
+    }
+
+    /**
+     * @param $currencyCodes
+     *
+     * @return array
+     */
+    private function _processActualRequest($currencyCodes)
+    {
+        $baseUrl =  'http://query.yahooapis.com';
+        $requestString = '/v1/public/yql?q=';
+        $query = 'select * from yahoo.finance.xchange where pair in ("'.implode('","', $currencyCodes).'")';
+        $query .= '&env=store://datatables.org/alltableswithkeys';
+        $query = str_replace(' ', '%20', $query);
+
+        $curl = curl_init();
+        curl_setopt($curl, CURLOPT_URL, $baseUrl . $requestString . $query);
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($curl, CURLOPT_HEADER, false);
+        $response = curl_exec($curl);
+        curl_close($curl);
+
+        $ratesXml = new \SimpleXMLElement($response);
+
+        $currencies = [];
+
+        /** @var \SimpleXMLElement $rate */
+        foreach ($ratesXml->results->rate as $rate) {
+            $id = (string)$rate->attributes()['id'];
+            $id = str_replace('=X', '', $id);
+
+            $currencies[$id] = (string)$rate->Rate;
+        }
+
+        return $currencies;
+    }
+
+    /**
+     * @param $currencyCodes
+     * @param \DateTime $date
+     *
+     * @return array
+     */
+    private function _processHistoricalRequest($currencyCodes, \DateTime $date)
+    {
+        $url = 'http://finance.yahoo.com/connection/currency-converter-cache';
```

Comment:
The problem is that yahoo.finance.historicaldata don't provide currencies info for Saturday and Sunday. And most important, it depends on US time zone. Due to the time difference, it will not provide information on the current day of at least 7 hours.

I suggest to use yahoo.finance.xchange for getting current day currencies. In another case, it is necessary to use some additional validation logic for date.